### DISCUSSION TITLE
regulator: regulator-fixed: device is never ready

### DISCUSSION THREAD

**Discussion by GrygiriiS** [zephyrproject-rtos/zephyr#73012] [Q&A] [answered]
It seems  regulator-fixed regulator device is never become ready (device_is_ready(reg_dev)) since one of the commits

- 539d4aa6e661 ("drivers: regulator: add common init enable API"
- b3fdb7477b11 ("drivers: regulator: fixed: Removed pin state setting from init")

and subsequent commits in drivers/regulator/regulator_common.c which added **regulator_common_init()** call to  the regulator-fixed itself, and non valid (for regulator-fixed)  **regulator_set_voltage()** calls in **regulator_common_init()**.

Is it know issue?

Hence,  **regulator_set_voltage()**  is not valid for regulator-fixed, and, actually, most of steps in **regulator_common_init()** - it seems reasonable to just call regulator_enable() for  regulator-fixed on init to update regulator state and ref_cnt.

Smth. like below:

```
--- a/drivers/regulator/regulator_fixed.c
+++ b/drivers/regulator/regulator_fixed.c
@@ -84,6 +84,7 @@ static const struct regulator_driver_api regulator_fixed_api = {
 static int regulator_fixed_init(const struct device *dev)
 {
        const struct regulator_fixed_config *cfg = dev->config;
+       int ret;
 
        regulator_common_data_init(dev);
 
@@ -93,14 +94,17 @@ static int regulator_fixed_init(const struct device *dev)
                        return -ENODEV;
                }
 
-               int ret = gpio_pin_configure_dt(&cfg->enable, GPIO_OUTPUT);
+               ret = gpio_pin_configure_dt(&cfg->enable, GPIO_OUTPUT);
 
                if (ret < 0) {
                        return ret;
                }
        }
 
-       return regulator_common_init(dev, false);
+       if (regulator_common_is_init_enabled(dev)) {
+               ret = regulator_enable(dev);
+       }
+       return ret;
 }

```

**Comment by edt-alex** [comment]
I'm looking at Zephyr v4.3.0-rc2 (commit e438b57a8ca) and I think the problem is that regulator_fixed.c does not provide a `.get_voltage` function in the DEVICE_API.

When the device tree sets min/max voltage of a fixed regulator, `regulator_common_init()` tries to read the voltage setting and adjust it to a valid range before possibly enabling the regulator. 

It's that `regulator_get_voltage()` call which is failing for me. I've locally patched regulator_fixed.c to add a get_voltage function (simply returns the `min-millivolts` value, like voltage list does). This seems to fix it for me.

**Reply by edt-alex** [reply, marked as answer]
https://github.com/zephyrproject-rtos/zephyr/issues/99339

**Comment by nandojve** [comment]
fixed by #99339
